src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -532,8 +543,14 @@ where
                         _ => false,
                     })
                     .await
-                    .unwrap();
-                prev_probe = (*new_probe).clone();
+                    .map(|probe| (*probe).clone())
+                    .unwrap_or_else(|_| {
+                        Some(Probe {
+                            probe_ts: (now_fn)().try_into().expect("must fit"),
+                            upstream_frontier: Antichain::new(),
+                        })
```

Comment:
what's the purpose of this `else` condition? Doesn't the `wait_for` closure guarantee that the value is a `Some`?

Reply:
It's for sources that terminate without producing a final probe for the empty frontier. If that happens we need to generate it here

Reply:
I think if we commit to using this new reclock strategy unconditionally we could make it such that sources are required to produce a probe before producing any data, which would force them to produce an empty probe before exiting